Clarify modelling tasks (#173)
hubverse-org/hubDocs: docs/source/user-guide/tasks.md

# Defining modeling tasks

> Every Hub is organized around "modeling tasks" that define how the target data (that is, the very thing the hub is trying to predict) should be modeled in terms of 

Reviewer: I got a little bit lost in the multiple phases of this sentence, maybe split it up into shorter pieces for more easy readability at the top of this page?

> 1. what factors to use for modelling (e.g. date, location, variant, etc) and 

Reviewer: "factors" feels to me like not quite the right word here. How about "variables"? I don't love it, but it feels a little bit more specific

2. the specific format of the model output.

The [tasks.json configuration file](#tasks-metadata)[^json] for a hub is used to structure the modelling tasks so that model submissions can be rapidly validated.
Modeling tasks are defined for single or multiple rounds[^multiround].
The three components of modeling tasks are:

 - The [`task_ids`{.codeitem} object](#task-id-vars) defines both labels for columns in submission files and the set of valid values for each column.
   Any unique combination of the values define a single modeling task, or target. 
 - The [`output_type`{.codeitem} object](#output-types) defines accepted representations _for each task_.
   More on the different output types can be found in [The Model Output Chapter](model-output.md#formats-of-model-output).
 - The [`target_metadata`{.codeitem} array](#target-metadata) provides additional information about each target.

> [^json]: Due to technical issues, we do not currently support json references or yaml metadata files.

Reviewer: what does it mean that we don't accept json references?

Author: ```suggestion
[^json]: Due to technical issues, we do not currently support [json references](http://niem.github.io/json/reference/json-schema/references/) or yaml metadata files.
```

(task-id-vars)=
## Task ID variables
Hubs typically specify that modeling outputs (e.g. forecasts or projections) should be generated for each combination of values across a set of task ID variables.<!-- 2024-09-01
Zhian: I am not sure if this 52-word sentence below adds anything.
It appears to be saying that model outputs can be evaluated for accuracy using truth data, which is also stated below.
For modeling exercises where the model outputs correspond to estimates or predictions of a quantity that could in principle be calculated from observable data, these task ID variables should be sufficient to uniquely identify an observed value for the modeling target that could be compared to model outputs to evaluate model accuracy.
This is discussed more in the section on [target (a.k.a. truth) data](#target-data).
-->
Because they are central to Hubs, task ID variables serve several purposes:

* **Define modelling tasks** of the hub in the hub metadata
* **Identify modelling tasks corresponding to forecasts** in the model outputs

> * **Allow alignment of model outputs with truth data** by specifying [target data](#target-data) that can help define methods to calculate outputs from "ground truth" data sources. 

Reviewer: I think it's confusing here how we talk about "truth data" and "target data". We've been trying to move away from "truth data" as a phrase, although talking about air-quotes "ground truth" is I think a helpful frame of reference. 

Maybe this could just read: 
**Allow alignment of model outputs with [target data](#target-data)** that are derived from "ground truth" data sources.

# Defining modeling tasks

> Every Hub is organized around "modeling tasks" that define how the target data should be modelled in terms of 

Author: If it's one thing I enjoy, it's repetition, redundancy, and repeating myself.

I say it's definitely worthwhile to include this.

> Every Hub is organized around "modeling tasks" that define how the target data should be modelled in terms of 

Author: ```suggestion
Every Hub is organized around "modeling tasks" that define how the target data (that is, the very thing the hub is trying to predict) should be modeled in terms of 
```

   Any unique combination of the values define a single modeling task, or target. 
 - The [`output_type`{.codeitem} object](#output-types) defines accepted representations _for each task_.
   More on the different output types can be found in [The Model Output Chapter](model-output.md#formats-of-model-output).
 - The [`target_metadata`{.codeitem} array](#target-metadata) provides additional information about each target.

[^json]: Due to technical issues, we do not currently support json references or yaml metadata files.

[^multiround]: For multiple rounds to share the same tasks without duplicating the `model_tasks` block, `round_id_from_variable` can be set to `true` and the `round_id` should be a column defined in the `task_ids`. See [the `tasks.json` schema](hub-config.md#hub-model-task-configuration-tasks-json-file) for details.

(task-id-vars)=
## Task ID variables
Hubs typically specify that modeling outputs (e.g. forecasts or projections) should be generated for each combination of values across a set of task ID variables.<!-- 2024-09-01

> Zhian: I am not sure if this 52-word sentence below adds anything.

Reviewer: agree

```

(task-id-use)=
### Usage of task ID variables

**Task ID variables represent columns in model output files.**
It's important to understand that model output files are in **tabular format (e.g. csv or parquet).**
Moreover, these tables are presented in a [long/narrow representation](https://en.wikipedia.org/wiki/Wide_and_narrow_data) where each row of data represents a unique combination of task ID variables and a single value from the model output[^tidy]. 

[^tidy]: This type of data is also known as "tidy data," which is a term coined by Hadley Wickham that's heavily used in the R community. You can read more about the concept in the [Data Tidying Chapter of the R4DS book](https://r4ds.hadley.nz/data-tidy#sec-tidy-data) and [the Tidy Data paper by Wickham (2014)](https://www.jstatsoft.org/article/view/v059i10).

#### Special Task ID variables

Task ID variables are used to parameterize modelling efforts. 
However, some task ID variables serve specific purposes to define submission rounds and targets.
Every hub must have **a single task ID variable that uniquely defines a submission round.**

> It has become a convention to use a task ID like `origin_date` or `forecast_date` for this purpose, although this is not a specific restriction.

Reviewer 2: **Update**: I might be wrong about the YYYY-MM-DD round_id format requirement. We had some discussions about it earlier this year, but I don't remember where we landed and can't really tell from our artifacts:
* https://github.com/orgs/hubverse-org/discussions?discussions_q=round_id
* https://github.com/hubverse-org/hubValidations/discussions/13

> It has become a convention to use a task ID like `origin_date` or `forecast_date` for this purpose, although this is not a specific restriction.

Author: Oh good point (which also relates to https://github.com/hubverse-org/hubDocs/pull/176#discussion_r1757531803).

The [parsing code in `hubValidations` still expects ISO-8601 dates](https://github.com/hubverse-org/hubValidations/blob/640f68083dc1e5246856bf9d103d838ed17154fc/R/parse_file_name.R#L33-L36) and I don't know that [the discussion in hubValidations](https://github.com/hubverse-org/hubValidations/discussions/13) has been resolved yet.

For example, in [Running Example 1](#running-examples), this task ID is `origin_date`.

There can be **one or more task ID variables to define a modelling "target"** (these are referred to in the [tasks metadata](#tasks-metadata) as a `target_key`).
For example, in our [Running Example 1](#running-example-1), the task ID variables are `target`, `location`, and `origin_date`.
In this example, `target` is the target key and can only take on one value "inc covid hosp".
In other examples, (such as [Running Example 3](#running-example-3)) more than one variable can serve as target keys together.
In example 3, both `outcome_variable` and `outcome_measure` make up the target keys.  

#### Proposed standard of Task ID variables

> In general, there are no restrictions on what task ID variables may be named, however when appropriate, we suggest that Hubs adopt the following standard task ID or column names and definitions[^new-vars]:  

Reviewer 3: The  main reason we encourage folks to use the naming we've specified alraady (if appropriate) is there is stronger validation of their config via the schema.

> In general, there are no restrictions on what task ID variables may be named, however when appropriate, we suggest that Hubs adopt the following standard task ID or column names and definitions[^new-vars]:  

Author: I updated this in 76f3f90 to mention the recommended vars first and then indicate that it's possible to add new ones, but with the caveat that they won't have the same validation advantage.

> * `origin_date`{.codeitem}: the starting point that can be used for calculating a `target_date` via the formula `target_date = origin_date + horizon * time_units_per_horizon` (e.g., with weekly data, `target_date` is calculated as `origin_date + horizon * 7` days).

Reviewer: 100% agree!

Although I will admit that I am sort of confused/distracted in this section by thinking about why some code is highlighted yellow and other is not. Even the same terms, like "target_date" sometimes is and sometimes isn't.

  For instance, if models are requested to forecast the number of hospitalizations that will occur on 2022-07-15, the `target_date` is 2022-07-15.
* `horizon`{.codeitem}: The difference between the `target_date` and the `origin_date` in time units specified by the hub (e.g., may be days, weeks, or months)
* `age_group`{.codeitem}: a unique identifier for an age group

[^new-vars]: As Hubs define new modeling tasks, they may need to introduce new task ID variables that have not been used before.
In those cases, the new variables should be added to this list to ensure that the concepts are documented in a central place and can be reused in future efforts.

(output-types)=
## Output types

The `output_type` object defines accepted representations for each task.
More on the different output types can be found in the [formats of model output section](#output-type-table) from the model output chapter.

(target-metadata)=
## Target metadata

Target metadata is an array in the [tasks.json](#tasks-metadata) schema file that defines the characteristics of each target. 
This serves as a logical connection between `task_ids` and corresponding `output_types`:

> ```{mermaid}

Reviewer: oooh, pretty. I've never used mermaid before, but it's a nice addition.

flowchart LR
    subgraph task-id["task_ids"]
    target
    end
    
    subgraph output-type["output_type"]
    vars["[output type objects]"]
    end

    subgraph target-metadata["target_metadata"]
    subgraph tk["target_keys"]
    tktarget["target"]
    end
    target-type["target_type"]
    end

    tktarget -->|"matches"| target
    target-type -->|"corresponds to"| vars
```

### Example
Here is an example of how the target metadata fields might appear in the `tasks.json` schema for a Hub whose target is incident covid hospitalizations. 

```
"target_metadata": [
    {
        "target_id": "inc covid hosp",
        "target_name": "Daily incident COVID hospitalizations",
        "target_units": "count",
        "target_keys": {
            "target": "inc covid hosp"
        },
        "description": "Daily newly reported hospitalizations where the patient has COVID, as reported by hospital facilities and aggregated in the HHS Protect data collection system.",
        "target_type": "discrete",
        "is_step_ahead": true,
        "time_unit": "day"
    }
]
```

### Details

> It is composed of the following fields:

Author: 💀!

> It is composed of the following fields:

Author: ```suggestion
Target metadata comprises the following fields:
```
